src/ConstructionBase.jl

```diff
@@ -72,5 +73,26 @@ function setproperties_unknown_field_error(obj, patch)
     throw(ArgumentError(msg))
 end
 
+function setproperties!(obj, patch::NamedTuple)
+    _setproperties!(obj, pairs(patch)...)
+end
+
+_setproperties!(obj) = obj
+
+function _setproperties!(obj, (key, val), tail...)
+    Base.setproperty!(obj, key, val)
+    _setproperties!(obj, tail...)
+end
+
+ismutablestruct(x) = ismutablestruct(typeof(x))
+Base.@pure ismutablestruct(T::DataType) = T.mutable
+
+function setproperties!!(obj, patch::NamedTuple)
+    if ismutablestruct(obj)
+        setproperties!(obj, patch)
+    else
+        setproperties(obj, patch)
+    end
+end
```

Comment:
OK, so `setproperties!!` I had in BangBang.jl haven't implemented the full functionalities I wanted to add.  I'm trying to do it in https://github.com/tkf/BangBang.jl/pull/23

Translating to `setproperties!!`, it would be something like this:

```julia
function setproperties!!(obj, patch::NamedTuple{pnames}) where pnames
    if ismutablestruct(obj) && all(n -> patch[n] isa fieldtype(typeof(obj), n), pnames)
        setproperties!(obj, patch)
    else
        setproperties(obj, patch)
    end
end
```